TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs

```diff
@@ -230,12 +231,39 @@ public virtual void StartCobolProgram(ProgramIdentification programIdentificatio
             Dispatcher.StartCobolProgram(programIdentification, libraryCopy);
         }
 
-        public virtual void EndCobolProgram(TypeCobol.Compiler.CodeElements.ProgramEnd end)
+        public virtual void EndCobolProgram(ProgramEnd end)
         {
-            AttachEndIfExists(end);
-            Exit();
-            programsStack.Pop();
-            Dispatcher.EndCobolProgram(end);
+            var endedProgramName = end?.ProgramName?.Name;
+            if (endedProgramName != null)
+            {
+                // Find the program in the stack that has the name specified by END PROGRAM
+                var matchingProgram = programsStack.FirstOrDefault(p => p.Name.Equals(endedProgramName, StringComparison.OrdinalIgnoreCase));
+                if (matchingProgram != null)
+                {
+                    // Pop all programs to have CurrentProgram as the program whose name match
+                    if (matchingProgram != CurrentProgram)
+                    {
+                        // Recursively close all nested programs
+                        EndCobolProgram(null);
+                    }
+                }
+                else
+                {
+                    if (CurrentProgram == null)
+                    {
+                        AttachEndIfExists(end);
+                    }
+                }
```

Comment:
I think this should be moved to the final if, superfluous `END PROGRAM` should be attached to root even if they are nameless. And could you please add a comment explaining that the end is attached to SourceFile and that this is checked later in a checker ?

Reply:
Nameless `END PROGRAM` should be correctly added to the SourceFile now.
A file to test this case has been added, it has the 3 new error cases.